### DISCUSSION TITLE
Unsuccessful convergence of a simple steady first order convection problem.

### DISCUSSION THREAD

**Discussion by Raven-pro** [idaholab/moose#21598] [Q&A General] [answered]
Hi, I am just getting started to use moose and I want to first solve a very simple steady, first order problem.
So here is a thing, I want to solve enthalpy of this equation, where W is set to be constant so i can put it out of the partial differential term.
![1657942422684](https://user-images.githubusercontent.com/68334741/179337651-b6ee1867-c303-4367-a356-90818cdf370a.jpg)
Then if I offer the boundary condition at the inlet, it can be easily solved using FDM and we can get the outlet value. However, in my moose application, if the boundary condition is set to the inlet 'left', the solve will not converge, while setting it to the outlet 'right', the calculation will proceed normally. 
But what's the difference? I cannot figure it out, can anybody help me with that? I'll appreciate it so much.
Below is my input file here.

```
[Mesh]
    type = GeneratedMesh
    dim = 1
    nx = 20
    xmax = 1.0
    xmin = 0
[]

[Variables]
    [H]
        order = FIRST
        family = LAGRANGE
    []
[]

[Kernels]
    [ker1]
        type = GradTerm
        variable = H
	velocity = '1.0 1.0 1.0'
    []
    [source1]
        type = ValTerm
        variable = H
    []
[]

[BCs]
    [bc1]
        type = DirichletBC
        variable = H
        value = 1133820.0
        boundary = "left"  
# If I set the boundary to "right", the solve will proceed normally  
    []
[]

[VectorPostprocessors]
  [H]
    type = NodalValueSampler
    variable = H
    #boundary = 'bottom'
    block = 0
    execute_on = 'initial timestep_end' 
    use_displaced_mesh = true
    sort_by = x
  []
[]

[Problem]
    type = FEProblem
    coord_type = XYZ
    rz_coord_axis = X
[]

[Executioner]
    type = Steady
    solve_type = 'NEWTON'
    automatic_scaling = true
[]

[Outputs]
    exodus = true
    csv = true
[]
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

So this is an advection problem. The information sees to be flowing in the positive X direction ( I m guessing based on the parameters, you have to look at how you implemented this )  in this 1D mesh. 
So the boundary condition has to be set at the inlet, which should be left not right. 

since this is actually the non-converging case, I would guess there is an issue in GradTerm

btw you dont need this parameter in XYZ
```
    rz_coord_axis = X
```

also you should initialize W. The initial guess of 0 might be difficult to solve from

Guillaume

**Reply by Raven-pro** [reply]
Thank you so much, I have revised my code, and tried to use JFNK method instead. The calculation has finally converged. Perhaps the JFNK method is better in performance than the Newton method after all.

**Reply by WilkAndy** [reply]
The fact that Newton is poor almost definitely implies that the Jacobian in your code is wrong.  Otherwise, Newton would be far superior.

**Reply by Raven-pro** [reply]
But I cannot find anything wrong in my code now, could you please have a look at them for me? THX!
[GradTerm.h](https://github.com/89018580/SingelPhase/blob/main/include/kernels/GradTerm%20(SFConflict%202020211251%40tsinghua.edu.cn%202022-07-17-13-19-45).h)
[ValTerm.h](https://github.com/89018580/SingelPhase/blob/main/include/kernels/ValTerm.h)
[GradTerm.C](https://github.com/89018580/SingelPhase/blob/main/src/kernels/GradTerm%20(SFConflict%202020211251%40tsinghua.edu.cn%202022-07-17-13-19-44).C)
[ValTerm.C](https://github.com/89018580/SingelPhase/blob/main/src/kernels/ValTerm%20(SFConflict%202020211251%40tsinghua.edu.cn%202022-07-17-13-19-44).C)
[input.i](https://github.com/89018580/SingelPhase/blob/main/problems/singel12.i)

**Reply by hugary1995** [reply]
1. In your GradTerm.C you have `_test[_qp][_i]` which is incorrect. It should be `_test[_i][_qp]` instead.
2. I agree with @WilkAndy 
3. ~I don't think the convergence issue is related to your BC.~
4. ~You have a pure advection problem, which _cannot_ be solved by FEM without stabilization.~

**Reply by WilkAndy** [reply]
Perhaps this is getting off-topic here, but i don't totally agree with @hugary1995 about advection: i believe FEM can solve advection without using stabilization, but in most cases (mesh sizes, timestep sizes) the errors are unacceptable.

Advection problems are also quite sensitive to correct boundary conditions.  Eg, it's quite easy to accidentally create a model which is trying to extract "stuff" (fluid, heat, etc) from a boundary when there actually isn't any "stuff" there to extract.  In that case MOOSE (and any decent numerical solver) will not converge because there truly isn't any solution.

**Reply by hugary1995** [reply]
Interesting... I thought you get instabilities from the inf-sup condition for advection-dominated advection-diffusion problem, which as you said, can be solved without stabilization (subject to acceptable errors).

~But it is different for pure advection problems. I haven't seen anyone solving it without stabilization IIRC. Let's say we look at the equation in the OP, there is no transient term, no artificial viscosity, no diffusion, no stabilization -- just advection and a constant source. Isn't the matrix singular?~

**Reply by hugary1995** [reply]
Ugh no, what was I taking about... there's a Dirichlet BC. So as long as he doesn't use -pc_type lu he should be fine.

**Reply by hugary1995** [reply]
And you are right about the BC as well.

**Reply by WilkAndy** [reply]
I think your comments raise an important point, @hugary1995  .   That is, @89018580 , are you sure that this is the problem you actually want to solve?  The solution is just H = linear function of x.  Do you also want some time dependence, @89018580  ?

**Reply by Raven-pro** [reply]
In fact I want to solve a more comlex problem where there are some time dependences. But as I am just getting started to use MOOSE, I think perhaps it's better to solve a simplified steaty problem as my first step. 
In the original  steady conservation equations, the time derivatives are removed, as there is no diffusion and viscosity, the velocity W can be seen as the constant.
<img width="742" alt="截屏2022-07-21 11 02 10" src="https://user-images.githubusercontent.com/68334741/180121150-aadd41fc-9b7f-4ca8-8c61-6cd77d921740.png">
The equations can be seen as coupling of variables P and H, in which H can be solved alone（rho is set as linear interpolation of H）. So I tried to solve H first and H should be a linear function of x. I thought it should be easy but I ran into the convergence problem.
So does that mean advection problems are not suitable to be solved with NEWTON?

**Reply by WilkAndy** [reply]
You really *want* to use NEWTON, because it will give far superior performance *if* your Jacobians are correct.  I don't know what your equations are really (eg, presumably the first equation means W is a real number) but if everything but P and H are simply real numbers then why don't you use the `ConservativeAdvection` Kernel that's already built into MOOSE, since it has a basic numerical stabilization and its Jacobian is definitely correct.

a

**Reply by Raven-pro** [reply]
> You really _want_ to use NEWTON, because it will give far superior performance _if_ your Jacobians are correct. I don't know what your equations are really (eg, presumably the first equation means W is a real number) but if everything but P and H are simply real numbers then why don't you use the `ConservativeAdvection` Kernel that's already built into MOOSE, since it has a basic numerical stabilization and its Jacobian is definitely correct.
> 
> a

THX！ I will try to use `ConservativeAdvection` instead.
